KAFKA-3452 Follow-up: Optimize ByteStore Scenarios

guozhangwang opened the pull request:
This is a refactoring follow-up of https://github.com/apache/kafka/pull/2166. Main refactoring changes:

1. Extract `InMemoryKeyValueStore` out of `InMemoryKeyValueStoreSupplier` and remove its duplicates in test package.

2. Add two abstract classes `AbstractKeyValueIterator` and `AbstractKeyValueStore` to collapse common functional logics.

3. Added specialized `BytesXXStore` to accommodate cases where key value types are Bytes / byte[] so that we can save calling the dummy serdes.

4. Make the key type in `ThreadCache` from byte[] to Bytes, as SessionStore / WindowStore's result serialized bytes are in the form of Bytes anyways, so that we can save unnecessary `Bytes.get()` and `Bytes.wrap(bytes)`.

Each of these should arguably be a separate PR and I apologize for the mess, this is because this branch was extracted from a rather large diff that has multiple refactoring mingled together and @dguy and myself have already put lots of efforts to break it down to a few separate PRs, and this is the only left-over work. Such PR won't happen in the future.

Ping @dguy @enothereska @mjsax for reviews

guozhangwang committed:
a first pass of the refactoring

guozhangwang committed:
refactor processor topology a bit to let processor manager be aware of logged stores beforehand

guozhangwang commented:
@dguy @mjsax for reviews (this is still WIP):

1. Follow the refactoring ideas in https://github.com/apache/kafka/pull/2166#pullrequestreview-15180711. Now the hierarchies can be viewed in `RocksDBKeyValueStoreSupplier`, `RocksDBWindowStoreSupplier`, `RocksDBSessionStoreSupplier`. Note that when caching or logging is enabled, all the underlying stores will be typed as `<Bytes, byte[]>` for performance optimizations. 

2. Added a few `AbstractXXX` classes to collapse common functions across some peer classes.

3. Refactored `Stores` a bit to better aligned with the supplier class.

4. Removed some redundant inner maps in `ProcessorTopology` which is only used `ProcessorStateManager` for flushing stores and remembering the changelog topic of the stores, instead let the `ProcessorTopology` to construct the full map from stores to their corresponding changelog topics if they are logging enabled within `build()`. It also allows to extract logging out of `RocksDB`.

Some more thoughts

5. With `WrapperStateStore`, we can now call `context.register(..)` in the outer-most store and call its `inner().put(...)` in the restoreCallback instead of calling it in the inner-most store, and hence can further get rid of the `root` parameter.

6. We can let users-customized aggregation / join operators to be per-store configurable so that we can also remove the `TupleForwarder` interface. I.e.:

```
StateStoreSupplier<...> mySupplier = new MyStateStoreSupplier(...)

// this means library will not add any wrappers for logging / caching / etc
// if users have caching mechanism in its only supplier, she needs to consider 
// flushing / eviction / forwarding herself.
stream.groupByKey(...).aggregate(..., supplier);   

// this means library will wrap user-provided underlying store engine as required
// users do not need to consider the above and our wrapper will do that without hacking forwarder.
StateStoreSupplier<...> myWrappedSupplier = Stores.create(storeName)
                .withKeys(keySerde)
                .withValues(aggValueSerde)
                .persistent()
                .enableCaching();
                .wrap(mySupplier);
                .build();

stream.groupByKey(...).aggregate(..., supplier);   
```

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -342,13 +240,51 @@ public StateStoreSupplier build() {
+        InMemoryWrapperFactory<K, V> disableLogging();
+    }
+
+    private static abstract class AbstractInMemoryFactory<K, V> implements InMemoryWrapperFactory<K, V> {
```
I'm not sure this is necessary? The only sub-class is `InMemoryKeyValueFactory`.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -360,27 +296,103 @@ public StateStoreSupplier build() {
+        PersistentWrapperFactory<K, V> disableCaching();
+    }
+
+    private static abstract class AbstractPersistentFactory<K, V> implements PersistentWrapperFactory<K, V> {
```
I don't think this should be an inheritance hierarchy. This class seems like it should be a standalone class that is used to collect the decisions/conifg etc for logging and caching. It could be constructed and passed in to what are currently its subclasses.
With inheritance we end up with competing interfaces in the hierarchy, i.e., the same methods on `PersistentWrapperFactory` and `PersistentKeyValueFactory`

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -389,50 +401,229 @@ public StateStoreSupplier build() {
+        }
+
+        @Override
+        public PersistentKeyValueFactory<K, V> enableLogging(final Map<String, String> config) {
```
This is an example of how the inheritance doesn't really work. It actually needs to return `this`. I think it would be much better to get rid of the inheritance and use delegation instead.

dguy commented:
@guozhangwang - i've made a quick pass to try and understand where this is heading. I left a couple of comments, but i largely decided not to comment too much as this is a WIP and i wasn't sure what you are after.

Anyway, i think the refactoring of the `StateStore` hierarchy generally looks good. The introduction of the `WrapperStateStore` interface and the abstract class makes the code a lot tidier. I'm not so sure about `AbstractSessionStore` and `AbstractWindowStore` as they don't offer much value, and i have a tendency to avoid unnecessarily deep inheritance (well, i'd actually prefer no impl inheritance at all!).

In `Stores` i think the addition of `PersistentSessionWindowedFactory` and `PersistentTimeWindowedFactory` are good. But i'm not really getting why we need `PersistentWrapperFactory` and the abstract classes that implement it? It seems to me that we just need a class that captures whether or not the stores are logged/cached etc. This could just be a shared class among all the factories that is created and passed to the constructor of the appropriate factory.

dguy commented:
Something i forgot to add. In `ProcessorStateManager` you have removed the below block of code in `flush`
```
final ProcessorNode processorNode = stateStoreProcessorNodeMap.get(store);
                if (processorNode != null) {
                    context.setCurrentNode(processorNode);
                }
```
It isn't clear to me where you intend for this logic to go now?

dguy commented:
Another thing. Point 6 you mention that we can get rid of `TupleForwarder`, yet i still don't see how that is possible? i.e, a user can still provide a store that is not cached - in which case we would still need it. 
Perhaps you are suggesting the `Store` or `WrappedStore` takes care of the forwarding?

guozhangwang commented:
About wappers in `Stores`: I think I agree. The original motivation was that we will let for example `XXXStoreSupplier.build()` to return `XXXStore` instead of not only `StateStore`, and hence in the DSL we will enforce `Window/SessionStoreSupplier` in aggregation / joins instead of just `StateStoreSupplier`, which actually allow users to pass a key-value store supplier that will only cause a RTE. But while implementing it I feel there maybe better ways to achieve this. So feel free to change.

About `stateStoreProcessorNodeMap`, my observation is that it is neither correct nor necessary: it links a store to the FIRST processor node that links to it, while multiple processors could link to this store, and it is used to set the `currentProcessorNode` in `context` when we call `store.flush`, but since we have already stored that information in dirtyEntry and always set/reset it in the flushing listener I feel it is not needed any more. Maybe I miss anything?

About `TupleForwarder`: my rationale is that, we can add some functions like `wrap` like I showed in the example to allow users to wrap their underlying storage engines with library-provided metering / logging / caching functionalities, AND if they don't do it, they either 1) do not really want caching or logging, which is fine; 2) they use their only logging / caching mechanism, in which case they should implement the `loggingEnabled` / etc functions to be false. Then in `KStream/KTableXXXAggregateProcessor.process()` implementation, we can decide if `windowStore isInstanceOf CachingWindowStore` to decide whether or not call `context.foward()` or not knowing that the caching store will execute this step itself; in this case we just won't do any coupling of dedupping with caching if user's customized store does that internally. And in the future, if we completely change KIP-63 to not coupling dedupping with caching but let user specifies triggers, then we can get rid of the `windowStore isInstanceOf` hacky logic as well.

@dguy BTW I think my refactoring is mingled with too many independent features and each of the above refactoring can be done in a single PR so we can make the reviewing process easier.

dguy commented:
w.r.t the `TupleForwarder` bit - the reason it exists at the moment is to avoid duplicating the decision etc of whether `context.forward()` should or shouldn't be called. I don't think this is actually changed by what is suggested, but maybe i'm just not seeing it at the moment. Anyway, i'll have a look and see where we get to

guozhangwang commented:
> w.r.t the TupleForwarder bit

I think you are right. As long as we do not have the full decision on which stores should always / never be caching enabled, we cannot drop it. I guess we can only revisit it if we consider reverting the unification of caching and dedupping.

dguy commented:
@guozhangwang - i've been trying to work with this, but it is an uphill battle. There are changes in here that definitely need a KIP, i.e, the refactoring in `Stores`, so that should be done separately. So i'd need to revert all those changes. Then i am having put a bunch of effort in to understand where you are heading, get it to compile, get tests to pass etc. I feel like i've been spinning my wheels and not getting anywhere.
So, i can't continue with this PR. I think the best course of action is to do this in smaller PRs. I.e, first do the StateStore hierarchy refactor. I can start that now. We need a KIP for the changes to `Stores`, so that is going to need to wait. There is also the `TopologyBuilder` etc stuff you were doing.

guozhangwang committed:
rebase on apache trunk

guozhangwang committed:
minor fixes

guozhangwang committed:
additional fixes

guozhangwang commented:
retest this please

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K3452-followup-state-store-refactor

guozhangwang committed:
rebase from apache trunk

guozhangwang committed:
minor fixes

guozhangwang committed:
rebase from apache trunk

guozhangwang committed:
make a pass to drop unnecessary changes

guozhangwang committed:
fix integration tests

guozhangwang committed:
last batch of unit test fixes

guozhangwang commented:
ping @dguy @mjsax @enothereska reviewing again.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -37,7 +37,7 @@
     protected final Set<String> sourceNodes;
 
-    public AbstractStream(KStreamBuilder topology, String name, Set<String> sourceNodes) {
+    AbstractStream(KStreamBuilder topology, String name, Set<String> sourceNodes) {
```
should we make the params `final` - seeing as we are here changing this line anyway

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -47,7 +47,7 @@ public AbstractStream(KStreamBuilder topology, String name, Set<String> sourceNo
     }
 
-    protected Set<String> ensureJoinableWith(AbstractStream<K> other) {
+    Set<String> ensureJoinableWith(AbstractStream<K> other) {
```
Here too

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -93,7 +93,7 @@ public void process(final K key, final V value) {
 
             try (final KeyValueIterator<Windowed<K>, T> iterator = store.findSessions(key, timestamp - windows.inactivityGap(),
-                                                                                      timestamp + windows.inactivityGap())) {
+                                                                                            timestamp + windows.inactivityGap())) {
```
what changed here?

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionKeySerde.java:
```diff
@@ -121,6 +120,13 @@ public static long extractStart(final byte [] binaryKey) {
         return ByteBuffer.wrap(binaryKey).getLong(binaryKey.length - TIMESTAMP_SIZE);
     }
 
+    public static Window extractWindow(final byte [] binaryKey) {
```
Can you add a test for this?

dguy replied:
Thanks for adding the test. One nit: looks like this method doesn't need to be public.

dguy commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionKeySerde.java:
```diff
@@ -129,10 +135,16 @@ public static long extractStart(final byte [] binaryKey) {
+        return new Windowed<>(key, window);
+    }
+
+    public static Windowed<Bytes> fromBytes(Bytes bytesKey) {
```
Would be good to have tests for this method and `from`, too.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractKeyValueIterator.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Override
+    public synchronized void close() {
```
why synchronized? The iterators are only ever called from a single thread, right?

guozhangwang replied:
Right.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractMergedSortedCacheStoreIterator.java:
```diff
@@ -32,14 +32,15 @@
  * @param <V>
  */
-abstract class AbstractMergedSortedCacheStoreIterator<K, KS, V> implements KeyValueIterator<K, V> {
+abstract class AbstractMergedSortedCacheStoreIterator<K, KS, V> extends AbstractKeyValueIterator<K, V> {
```
I'm not convinced we gain anything by adding another level of inheritance here. It removes one method that doesn't do much anyway. My preference would be to leave it as it was. If we are going to use inheritance then i think we should keep the hierarchy as shallow as possible.

guozhangwang replied:
agree.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractStoreSupplier.java:
```diff
@@ -24,15 +24,15 @@
     protected final Map<String, String> logConfig;
 
-    public AbstractStoreSupplier(String name, Serde<K> keySerde, Serde<V> valueSerde, Time time, boolean logged, Map<String, String> logConfig) {
+    AbstractStoreSupplier(String name, Serde<K> keySerde, Serde<V> valueSerde, Time time, boolean logged, Map<String, String> logConfig) {
```
I know this was here already, but seeing as you are changing this line might as well make the args final

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:
```diff
@@ -63,14 +64,16 @@ public void init(final ProcessorContext context, final StateStore root) {
         this.name = context.taskId() + "-" + underlying.name();
         this.cache = this.context.getCache();
+
+        // set cache flush listener
```
Do we need the comment? It seems pretty obvious to me what it is doing

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:
```diff
@@ -119,30 +120,35 @@ public synchronized void flush() {
-        byte[] binaryFrom = WindowStoreUtils.toBinaryKey(key, timeFrom, 0, serdes);
-        byte[] binaryTo = WindowStoreUtils.toBinaryKey(key, timeTo, 0, serdes);
+
+        Bytes binaryFrom = Bytes.wrap(WindowStoreUtils.toBinaryKey(key, timeFrom, 0, serdes));
```
Perhaps change `WindowStoreUtils.toBinaryKey(...)` to return `Bytes` - then we wouldn't need to do the `Bytes.wrap` everywhere?

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,190 @@
+
+    private StateSerdes<K, V> serdes;
+
+    InMemoryKeyValueStore(String name, Serde<K> keySerde, Serde<V> valueSerde) {
```
would be nice to mark all params final whilst we are changing this

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MergedSortedCacheKeyValueStoreIterator.java:
```diff
@@ -36,7 +36,7 @@
 
     @Override
-    public KeyValue<K, V> deserializeStorePair(KeyValue<Bytes, byte[]> pair) {
+    public KeyValue<K, V> deserializeStorePair(final KeyValue<Bytes, byte[]> pair) {
```
:-)

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java:
```diff
@@ -44,14 +45,53 @@
+        }
+    }
+
+    static RocksDBSessionStore<Bytes, byte[]> bytesStore(final SegmentedBytesStore inner) {
```
This method doesn't appear to be used anywhere?

guozhangwang replied:
In fact this is because the RocksDBStoreSupplier should trigger this function but it does not. Have fixed it.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java:
```diff
@@ -34,31 +34,51 @@
-        return new RocksDBWindowStore<>(inner, Serdes.Bytes(), Serdes.ByteArray(), retainDuplicates);
+        @Override
+        public void put(Bytes key, byte[] value, long timestamp) {
+            if (retainDuplicates) {
```
maybe extract the retain duplicates check into a method on the super class?

guozhangwang replied:
ack.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStore.java:
```diff
@@ -106,28 +126,56 @@ public void put(K key, V value, long timestamp) {
 
-    private static class TheWindowStoreIterator<V> implements WindowStoreIterator<V> {
-        private final KeyValueIterator<Bytes, byte[]> actual;
+    // TODO: this iterator may need to be extracted out of RocksDBWindowStore to be used for other underlying window store implementation
```
Perhaps just extract it? Would have the added benefit of being able to test these classes in isolation of RocksDB

guozhangwang replied:
ack.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreUtils.java:
```diff
@@ -32,12 +32,11 @@
 
-
-    public static <K> byte[] toBinaryKey(K key, final long timestamp, final int seqnum, StateSerdes<K, ?> serdes) {
+    static <K> byte[] toBinaryKey(K key, final long timestamp, final int seqnum, StateSerdes<K, ?> serdes) {
```
As mentioned above - perhaps change this to return `Bytes` and then we won't need to `Bytes.wrap` in various places

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WrapperKeyValueStore.java:
```diff
@@ -0,0 +1,40 @@
+ */
+interface WrapperKeyValueStore<K, V> extends WrapperStateStore, KeyValueStore<K, V> {
+
+    abstract class AbstractKeyValueStore<K, V> extends WrapperStateStore.AbstractStateStore implements WrapperKeyValueStore<K, V> {
```
I'm not really convinced of the benefit of this abstract class

guozhangwang replied:
agree.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WrapperStateStore.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * A storage engine wrapper for utilities like logging, caching, and metering.
+ */
+interface WrapperStateStore extends StateStore {
```
This looks like duplicate of `WrappedStateStore`

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K3452-followup-state-store-refactor

guozhangwang committed:
github comments v1

guozhangwang committed:
github comments two

dguy commented:
build is failing with compilation error:
>scala2.10/streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java:155: error: method wrap in class Bytes cannot be applied to given types;
>      underlying.put(Bytes.wrap(WindowStoreUtils.toBinaryKey(key, DEFAULT_TIMESTAMP, 0, >WindowStoreUtils.INNER_SERDES)), "a".getBytes());
>                          ^
>required: byte[]
>found: Bytes

guozhangwang commented:
> build is failing with compilation error:

Thanks for pointing this out, I was not done addressing all comments nor completed the implementation yet last night, will continue today.

guozhangwang committed:
github comments v3

guozhangwang committed:
extracted wrapper iterators

guozhangwang committed:
address github comments, v4

guozhangwang commented:
@dguy ready for review again. Some changes:

1. Refactored RocksDBWindowStoreTest a bit to put common variables as class variables.
2. Use the byte stores in RocksDBSessionStoreSupplier.

dguy commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -73,53 +99,27 @@
+    @Override
     public void remove(final Windowed<K> sessionKey) {
         validateStoreOpen();
+
```
nit: any reason for the extra line?

guozhangwang replied:
re `One nit: looks like this method doesn't need to be public.` Actually the test needs it to be package-private, and while on that, I feel we can either make it pure private and remove the corresponding test, or leave it public with the test if we speculate it to be used in the future.

I'd go with the second option if you do not have a strong opinion.

dguy commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java:
```diff
@@ -59,56 +57,55 @@
-    private static final long DEFAULT_CACHE_SIZE_BYTES = 1 * 1024 * 1024L;
-    private ThreadCache cache;
     private final Segments segments = new Segments(windowName, retentionPeriod, numSegments);
+    private final StateSerdes<Integer, String> serdes = new StateSerdes<>("", Serdes.Integer(), Serdes.String());
```
Nice tidy up of this test class :-)

dguy reviewed (commented):
Overall LGTM. I left a couple of very minor nit picks.
Also, should we add tests for `WrappedWindowStoreIterator` and `WrappedSessionStoreIterator` ?

guozhangwang commented:
> Also, should we add tests for WrappedWindowStoreIterator and WrappedSessionStoreIterator ?

I tried to add some tests, but while doing that I found all the logic is covered in corresponding unit tests of `SessionKeySerde` and `WindowStoreUtils`, so I decided to not introduce overlap coverage.

guozhangwang committed:
minor fixes

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/DelegatingPeekingKeyValueIterator.java:
```diff
@@ -76,7 +80,7 @@ public synchronized boolean hasNext() {
     @Override
     public void remove() {
-        throw new UnsupportedOperationException("remove not supported");
+        throw new UnsupportedOperationException("remove() is not supported in DelegatingPeekingKeyValueIterator.");
```
Nit: `"... DelegatingPeekingKeyValueIterator"` -> `"..." + getClass().getName()`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:
```diff
@@ -67,7 +67,7 @@ public void init(final ProcessorContext context, final StateStore root) {
-    void initInternal(final ProcessorContext context) {
+    private void initInternal(final ProcessorContext context) {
         this.context = (InternalProcessorContext) context;
         this.serdes = new StateSerdes<>(underlying.name(),
```
Nit: remove `this`
(we should avoid `this` whenever possible)

guozhangwang replied:
I'd prefer to leave it here for better code reading.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,190 @@
+        this.map = new TreeMap<>();
+    }
+
+    InMemoryKeyValueStore(String name) {
```
add `final` (also all other methods below)

guozhangwang replied:
`KeyValueStore` interface does not mark all the parameters as final, and I think it is safer to keep it as is. As for this constructor, will remove it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,190 @@
+
+    @Override
+    public void close() {
+        this.open = false;
```
Should we do `map.clear()` here?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,190 @@
+
+        @Override
+        public void close() {
+            // do nothing
```
May set `iter = null` to ensure this iterator is not use after closing anymore?

guozhangwang replied:
This iterator is always wrapped in `DelegatingPeekingKeyValueIterator`, which already does the isOpen validation.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java:
```diff
@@ -0,0 +1,190 @@
+
+        @Override
+        public K peekNextKey() {
+            throw new UnsupportedOperationException("peekNextKey() not supported on InMemoryKeyValueIterator");
```
"InMemoryKeyValueIterator" -> `getClass().getName()

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentIterator.java:
```diff
@@ -86,7 +87,6 @@ public boolean hasNext() {
 
     public void remove() {
-        throw new UnsupportedOperationException("remove not supported");
+        throw new UnsupportedOperationException("remove() is not supported in SegmentIterator");
```
getClass().getName()

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedSessionStoreIterator.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("remove() is not supported in WrappedSessionStoreIterator.");
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedWindowStoreIterator.java:
```diff
@@ -0,0 +1,93 @@
+
+        @Override
+        public KeyValue<Long, byte[]> next() {
+            if (!bytesIterator.hasNext()) {
```
Do we need this? Should `bytesIterator.next()` not throw the same exception for this case?

guozhangwang replied:
Since the underlying iterator could be provided by users which does not validate in its own `next()` call, it's better doing such validation in the higher-level where the library controls.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedWindowStoreIterator.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("remove() is not supported in WrappedWindowStoreIterator.");
```
as above.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueStoreTest.java:
```diff
@@ -174,7 +173,7 @@ public void shouldQueryRange() throws Exception {
+        final KeyValueIterator<String, String> range = store.range(hello, "zooom");
         assertThat(range.next(), equalTo(KeyValue.pair(hello, world)));
         assertThat(range.next(), equalTo(KeyValue.pair(hi, there)));
         assertFalse(range.hasNext());
```
I am confused. If we change `store.range(hello, "zooom");` should `hasNext()` not return `true` (start and end are inclusive). Thus, to me it seems the test is wrong and there must be a bug in the iterator code.

guozhangwang replied:
Great catch!

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/ReadOnlyWindowStoreStub.java:
```diff
@@ -106,7 +106,7 @@ public void close() {
         @Override
         public Long peekNextKey() {
-            throw new UnsupportedOperationException("peekNextKey not supported in stub");
+            throw new UnsupportedOperationException("peekNextKey() not supported in ReadOnlyWindowStoreStub.");
```
as above

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreTest.java:
```diff
@@ -59,56 +57,55 @@
     public void setUp() {
-        cache = new ThreadCache("testCache", DEFAULT_CACHE_SIZE_BYTES, new MockStreamsMetrics(new Metrics()));
+        baseDir = TestUtils.tempDirectory("test");
+        context = new MockProcessorContext(baseDir, Serdes.ByteArray(), Serdes.ByteArray(), recordCollector, cache);
```
Can't we initialize both vars directly and remove this method?

guozhangwang committed:
github comments

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/ChangeLoggingKeyValueStoreTest.java:
```diff
@@ -173,7 +173,7 @@ public void shouldQueryRange() throws Exception {
         store.put(hi, there);
         store.put("zooom", "home");
-        final KeyValueIterator<String, String> range = store.range(hello, "zooom");
+        final KeyValueIterator<String, String> range = store.range(hello, hi);
```
I am still puzzled why the test did not fail with "zooom" as end range. Can you explain? \cc @dguy WDYT?

guozhangwang replied:
Because in the old code of `InMemoryKeyValueStore`, our range function is implemented as `subMap(from, true, to, false)` while I changed it to `(from, true, to, true)` in the latest commit.

mjsax replied:
Ack. Overlooked this change :)

mjsax reviewed (approved):
LGTM

junrao commented:
Thanks for the patch. LGTM

guozhangwang commented:
Merged to trunk.
